topdown/object.go

```diff
@@ -32,21 +32,30 @@ func builtinObjectUnionN(_ BuiltinContext, operands []*ast.Term, iter func(*ast.
 		return err
 	}
 
-	r := ast.NewObject()
-	err = arr.Iter(func(t *ast.Term) error {
-		var o ast.Object
-		o, err = builtins.ObjectOperand(t.Value, 1)
+	// Because we need merge-with-overwrite behavior, we can iterate
+	// back-to-front, and get a mostly correct set of key assignments that
+	// give us the "last assignment wins, with merges" behavior we want.
+	// However, if a non-object overwrites an object value anywhere in the
+	// chain of assignments for a key, we have to "freeze" that key to
+	// prevent accidentally picking up nested objects that could merge with
+	// it from earlier in the input array.
+	// Example:
+	//   Input: [{"a": {"b": 2}}, {"a": 4}, {"a": {"c": 3}}]
+	//   Want Output: {"a": {"c": 3}}
+	result := ast.NewObject()
+	frozenKeys := map[*ast.Term]struct{}{}
+	for i := arr.Len() - 1; i >= 0; i-- {
+		o, err := builtins.ObjectOperand(arr.Elem(i).Value, 1)
```

Comment:
[nit] not new code, but I think this error will be slight off the mark. If it we're passing `[{"foo":"2"}, 1]`, it'll complain about operand 1 not being an object, I think.

Reply:
As I understand it, this builtin is supposed to take in an array of objects. The example input `[{"foo":"2"}, 1]` indeed should be rejected, right?

Reply:
It should but the error will say something about requiring an object operand 1, not an array of objects.

Reply:
Ah, I see. I'll see what I can do to render a better error message there.

Reply:
`builtins.NewOperandElementErr` looks to be pretty close to what we want on a failed type assertion here, I think?

Reply:
💡 I wasn't even aware that this existed. Great. Yes!